presto-hive/src/main/java/com/facebook/presto/hive/GenericHiveRecordCursor.java

```diff
@@ -309,7 +310,10 @@ private static long getLongExpressedValue(Object value, DateTimeZone hiveTimeZon
             long hiveMillis = JVM_TIME_ZONE.convertUTCToLocal(parsedJvmMillis);
 
             // convert to UTC using the real time zone for the underlying data
-            return hiveTimeZone.convertLocalToUTC(hiveMillis, false);
+            long utcMillis = hiveTimeZone.convertLocalToUTC(hiveMillis, false);
+
+            long micros = (((Timestamp) value).getNanos() / 1000) % 1000;
+            return millisToMicros(utcMillis) + micros;
```

Comment:
This has a bug when thee utcMilllis is negative. Modulo will yield a negative number and they are hard to verify. All of them should use the standard java converter or be isolated to a utility class with good unit tests.